driver-core/src/main/com/mongodb/internal/authentication/AwsInternalCredential.java

```diff
@@ -0,0 +1,211 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.internal.authentication;
+
+import com.mongodb.AwsCredential;
+import com.mongodb.MongoInternalException;
+import com.mongodb.lang.NonNull;
+import com.mongodb.lang.Nullable;
+import org.bson.BsonDocument;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Supplier;
+
+import static com.mongodb.assertions.Assertions.assertNotNull;
+import static com.mongodb.assertions.Assertions.isTrueArgument;
+
+public final class AwsInternalCredential {
+    private final String accessKeyId;
+    private final String secretAccessKey;
+    private final String sessionToken;
+
+    /**
+     * Construct a new instance.
+     *
+     * @param accessKeyId     the non-null access key ID that identifies the temporary security credentials.
+     * @param secretAccessKey the non-null secret access key that can be used to sign requests
+     * @param sessionToken    the non-null session token
+     */
+    public AwsInternalCredential(@Nullable final String accessKeyId, @Nullable final String secretAccessKey,
+            @Nullable final String sessionToken) {
+        this.accessKeyId = accessKeyId;
+        this.secretAccessKey = secretAccessKey;
+        this.sessionToken = sessionToken;
+    }
+
+    /**
+     * Gets the access key ID that identifies the temporary security credentials.
+     *
+     * @return the accessKeyId, which may not be null
+     */
+    @Nullable
+    public String getAccessKeyId() {
+        return accessKeyId;
+    }
+
+    /**
+     * Gets the secret access key that can be used to sign requests.
+     *
+     * @return the secretAccessKey, which may not be null
+     */
+    @Nullable
+    public String getSecretAccessKey() {
+        return secretAccessKey;
+    }
+
+    /**
+     * Gets the session token.
+     *
+     * @return the sessionToken, which may not be null
+     */
+    @Nullable
+    public String getSessionToken() {
+        return sessionToken;
+    }
+
+    public static AwsInternalCredential obtainFromEnvironment(final AwsInternalCredential startingCredential,
+            final Supplier<AwsCredential> awsCredentialSupplier) {
+
+        if (awsCredentialSupplier != null) {
+            return obtainFromSupplier(startingCredential, awsCredentialSupplier);
+        }
+
+        BsonDocument ec2OrEcsResponse = null;
+
+        String accessKeyId;
+        if (startingCredential.getAccessKeyId() != null) {
```

Comment:
A style difference is that I do not build the aws credential in pieces. If a given credential source has the credentials, I get it all or nothing from a given source. Your code allows one to pass `accessKeyId` in `startingCredential` and `secretAcccessKeyId` from the environment variable.

I think of the credential as a single thing, not a thing that can be pieced together from different sources.
_Example pseudo code_
```
if (startingCredential != null) {
  return startCredential
if System.getenv("AWS_ACCESS_KEY_ID") != null {
  .. get other env
  return
}
Do the getEc2OrEcsResponse stuff
...
```
This is how the server code checks credentials - https://github.com/mongodb/mongo/blob/c92b95b22c3290a6235f7952c598c8c5feae669e/src/mongo/client/sasl_aws_client_conversation.cpp#L94-L117

Reply:
Yeah, this definitely confused me too.  But for this PR I was trying to not change any existing behavior, and just refactor.  And I think this is also how the driver currently behaves.

Reply:
If it's wrong, I will open a separate ticket for that.

Reply:
Sounds good